### ISSUE TITLE
[VMStorage] data over retentionPeriod 1d is not removed

### ISSUE THREAD

**Issue by tantm3** [VictoriaMetrics/VictoriaMetrics#4544]
### Is your question request related to a specific component?

VictoriaMetrics cluster version v1.87.6
VMStorage component

### Describe the question in detail

I am using VictoriaMetrics cluster version with all components replica = 3
The `retentionPeriod=1d` is passed to VMStorage component, but it does not work as expected.
At the time I post this issue 2023:06:30 10:20, Grafana dashboard still shows data from 2023-06-28 16:15:00 that is `I think outside of 1 day period ( 2023:06:29 10:20 => 2023:06:30 10:20)`

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/7d49801f-9384-4ea3-ac43-37b7209bc3de)

I read through the session about retention here: (https://docs.victoriametrics.com/#retention), but I am confused about when vmstorage will remove old data with day retention, as the documentation only talks about monthly rotate data
```
Data is split in per-month partitions inside <-storageDataPath>/data/{small,big} folders. Data partitions outside the configured retention are deleted on the first day of the new month
```
Here is all config I have set in the VMStorage component
```
        - '--retentionPeriod=1d'
        - '--storageDataPath=/storage'
        - '--envflag.enable=true'
        - '--envflag.prefix=VM_'
        - '--loggerFormat=json'
        
```
Thanks for your time reading my question!

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [X] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by tantm3** [comment]
Additional information may help
- I had a VictoriaMetrics upgrade version from v1.83.0 => to v1.87.6 20 days ago
- I also go through this issue: https://github.com/VictoriaMetrics/VictoriaMetrics/issues/3502 and try to reset the cache at the VMStorage node, but it's seem not worked

**Comment by hagen1778** [comment]
Hello!

> Please note, the time range covered by data part is not limited by retention period unit. Hence, data part may contain data for multiple days and will be deleted only when fully outside the configured retention.

see https://docs.victoriametrics.com/#retention

**Comment by tantm3** [comment]
Hi @hagen1778, 

I am still confused after reading documentation about the retention.
Could you please explain to me using the times from the attached Grafana image?

**Comment by hagen1778** [comment]
I've tried to make it more clear here https://github.com/VictoriaMetrics/VictoriaMetrics/pull/4568
Let me know if there are suggestions to make it better.

In short, one data part can exceed the retention unit. For example, it is allowed to have data part to coved time range > 1d. Hence, this data part can be deleted only when fully out of retention period. For example, if data part covers 30h of data and your retentionPeriod=24h - then this data part will be deleted only after 30h, not 24h.

**Comment by tantm3** [comment]
I think it's a good idea to add the mention of `parts and partitions`.
But, could you give some scenarios that lead to the scenario `data part may contain data outside of retention period unit`?

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/3ba014f9-8704-4b31-82ba-bd60624cb772)

At the moment, the data from my VMStorage components are still outside the 1 day period, and I am pretty sure that I am not ingesting old timestamp metrics.

**Comment by hagen1778** [comment]
> At the moment, the data from my VMStorage components are still outside the 1 day period, and I am pretty sure that I am not ingesting old timestamp metrics.

Metrics with timestamps older than the retention period are rejected immediately, so they won't appear in database.

I'd recommend checking your storagePath and see if there are data parts with timestamps within retention period:
<img width="752" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/2902918/cb2767b2-b7a7-45f7-8f2d-d03a70e33778">
In my example, one data part covers almost 4d of data. If I'd set `retentionPeriod=1d` now this part will be deleted only when `MaxTimestamp` will become older than `1d`.

If that's not the case for you unexpected query results, I'd recommend checking https://docs.victoriametrics.com/Troubleshooting.html#unexpected-query-results
It might be that you're seeing just a cached response.

**Comment by tantm3** [comment]
> I'd recommend checking your storagePath and see if there are data parts with timestamps within retention period

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/d7b8bd6e-8857-4209-92c3-f7984cfcc2a8)

It's weird that the `metadata.json` file does not exist in any data parts.

**Comment by tantm3** [comment]
Here is the full query trace: [trace.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/11953964/trace.txt)
I tried to trace a query and I found out that in the `search for metricIDs in the previous indexdb` phase it's found and return 1 metrics id from day `2023-07-04`.
```
{
"duration_msec": 21.362,
"message": "rpc at vmstorage vm-1d-victoria-metrics-cluster-vmstorage-1.vm-1d-victoria-metrics-cluster-vmstorage.vmonitor-platform-backend.svc.cluster.local:8401",
"children": [
{
"duration_msec": 20.824,
"message": "rpc call search_v7()",
"children": [
{
"duration_msec": 16.175,
"message": "vmstorage-20230518-231341-tags-v1.87.6-cluster-0-g468cd5f7b: rpc call search_v7() at vmstorage",
"children": [
{
"duration_msec": 16.004,
"message": "init series search: filters=[{AccountID=0,ProjectID=3214014564,__name__=\"vdb.cpu.idle_perc\",database_name=\"vccvcvvccv\"}], timeRange=[2023-07-03T04:00:00Z..2023-07-05T04:05:00Z]",
"children": [
{
"duration_msec": 6.049,
"message": "search for matching metricIDs: filters=[{AccountID=0,ProjectID=3214014564,__name__=\"vdb.cpu.idle_perc\",database_name=\"vccvcvvccv\"}], timeRange=[2023-07-03T04:00:00Z..2023-07-05T04:05:00Z]",
"children": [
....
{
"duration_msec": 3.652,
"message": "search for metricIDs in the previous indexdb",
"children": [
  {
    "duration_msec": 0.003,
    "message": "search for metricIDs in tag filters cache",
    "children": [
      {
        "duration_msec": 0.003,
        "message": "cache miss"
      }
    ]
  },
  {
    "duration_msec": 1.591,
    "message": "search for metric ids: filters=[{AccountID=0,ProjectID=3214014564,composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}], timeRange=[2023-07-03T04:00:00Z..2023-07-05T04:05:00Z], maxMetrics=300000",
    "children": [
      {
        "duration_msec": 1.551,
        "message": "update metric ids: filters={AccountID=0,ProjectID=3214014564,composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}, timeRange=[2023-07-03T04:00:00Z..2023-07-05T04:05:00Z]: updated 1 metric ids",
        "children": [
          {
            "duration_msec": 1.544,
            "message": "parallel search for metric ids in per-day index: filters={AccountID=0,ProjectID=3214014564,composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}, dayRange=[19541..19543]",
            "children": [
              {
                "duration_msec": 0.695,
                "message": "parallel thread for date=2023-07-03",
                "children": [
                  .....
                ]
              },
              {
                "duration_msec": 1.532,
                "message": "parallel thread for date=2023-07-04",
                "children": [
                  {
                    "duration_msec": 0.343,
                    "message": "search for metric ids on a particular day: filters={AccountID=0,ProjectID=3214014564,composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}, date=2023-07-04, maxMetrics=300001",
                    "children": [
                      {
                        "duration_msec": 0.338,
                        "message": "search for the first non-negative filter with the smallest cost",
                        "children": [
                          {
                            "duration_msec": 0.324,
                            "message": "get metric ids for filter and date: filter={composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}, date=2023-07-04, maxMetrics=15000050, maxLoopsCount=9223372036854775807",
                            "children": [
                              {
                                "duration_msec": 0.319,
                                "message": "found 1 metric ids for filter={composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"} using exact search; spent 2 loops"
                              }
                            ]
                          },
                          {
                            "duration_msec": 0.009,
                            "message": "the filter={composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"} matches less than 15000050 series (actually 1 series); use it"
                          }
                        ]
                      },
                      {
                        "duration_msec": 0,
                        "message": "intersect the remaining 0 filters with the found 1 metric ids"
                      },
                      {
                        "duration_msec": 0,
                        "message": "found 1 metric ids"
                      }
                    ]
                  }
                ]
              },
              {
                "duration_msec": 0.543,
                "message": "parallel thread for date=2023-07-05",
                "children": [
                  {
                    "duration_msec": 0.503,
                    "message": "search for metric ids on a particular day: filters={AccountID=0,ProjectID=3214014564,composite(vdb.cpu.idle_perc,database_name)=\"vccvcvvccv\"}, date=2023-07-05, maxMetrics=300001",
                    "children": [
                      ...
                    ]
                  }
                ]
              }
            ]
          }
        ]
      }
    ]
  },
  {
    "duration_msec": 0,
    "message": "sort 1 matching metric ids"
  },
  {
    "duration_msec": 0.016,
    "message": "put 1 metricIDs in cache",
    "children": [
      {
        "duration_msec": 0,
        "message": "marshaled 1 metricIDs into 16 bytes"
      },
      {
        "duration_msec": 0.014,
        "message": "stored 1 metricIDs into cache"
      }
    ]
  }
]

```

**Comment by hagen1778** [comment]
> I think it's because I run VictoriaMetrics in Cluster mode and enable replicaFactor=2 in the vminsert component.

No, this means you run VictoriaMetrics version older than 1.90. Before 1.90 part names contain the metainformation from metadata.json. If you check part names you'll see they contain number-of-rows_number-of-blocs_start-date_end-date. According to the screenshot you have one part which stores data for 04 and 05 of July. This part will be possible to delete only on 6th of July. Another part stores data for 04 July. 

Data deletion for retentionPeriods < 1mo doesn't work on schedule. It happens during [background merges](https://medium.com/@valyala/how-victoriametrics-makes-instant-snapshots-for-multi-terabyte-time-series-data-e1f3fb0e0282). That's why doc says:
> Data parts outside the configured retention are eventually deleted during [background merge](https://medium.com/@valyala/how-victoriametrics-makes-instant-snapshots-for-multi-terabyte-time-series-data-e1f3fb0e0282).

You can force the deletion via [forced merge call](https://docs.victoriametrics.com/#forced-merge). 

Have you eliminated p3 from https://docs.victoriametrics.com/Troubleshooting.html#unexpected-query-results ?

**Comment by tantm3** [comment]
After I did [forced merge call](https://docs.victoriametrics.com/#forced-merge). and removed cached data using `/internal/resetRollupResultCache`, the data in date `04-07-2023` is removed:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/ffe2fd2f-1c89-4aea-94f9-f8741c383b75)

> Have you eliminated p3 from https://docs.victoriametrics.com/Troubleshooting.html#unexpected-query-results ?
To be honest, I read through that before, but I could not find out the one that help me understand the problem, I will read it again.

The last thing I am still confused about is why there is a `part stores data for 04 July` which is out of retention.

Could you please give me and everyone reading this issue some circumstances that may lead to that problem?

**Comment by hagen1778** [comment]
> The last thing I am still confused about is why there is a part stores data for 04 July which is out of retention.

VM stores retentionPeriod+1 period of data - see https://docs.victoriametrics.com/#retention. Also, deletion of data parts (unlike partitions) happens eventually (not on schedule).

See more details about retention periods lower than 1mo here https://github.com/VictoriaMetrics/VictoriaMetrics/issues/173

**Comment by tantm3** [comment]
Thanks for your help!
I will close this issue here because it has been solved.

**Comment by tantm3** [comment]
Hi @hagen1778,

I am sorry because of reopening the issue. 
But, It seems like the [background merge ](https://medium.com/@valyala/how-victoriametrics-makes-instant-snapshots-for-multi-terabyte-time-series-data-e1f3fb0e0282) is not running as expected.

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/18bb8557-843f-40ee-b0aa-f681887186f4)

Here is the result when I query the data from VictoriaMetrics Cluster with 1-day retention.
As we discussed earlier `VM stores retentionPeriod+1 period of data`, but I am confused why data is missing in a few time periods (in the image above).
And the deletion of the `part` in VM, Could we have a deep dive session to explain the `MaxTimestamp` in one `part` data?
I read through this block of code here and understand `a data part can be deleted only when fully outside the configured retention`
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/12874d0c-62b7-4c4b-93a5-44bb0e751679)

But, I am not sure what controls the max time range in one `part` data.
If the `MaxTimestamp` in one `part` data is keeping growing, that `part` is not removed, so I think we need to understand that.

**Comment by hagen1778** [comment]
Hello @tantm3!

> Here is the result when I query the data from VictoriaMetrics Cluster with 1-day retention.

Are you sure those gaps are caused by the retention and not because the metric was just absent at that time?

> If the MaxTimestamp in one part data is keeping growing, that part is not removed, so I think we need to understand that.

MaxTimestamp can't keep growing for a part because parts are immutable: once part is created, it can only be deleted afterwards. Multiple parts can be merged into the bigger part, and this new part will cover min MinTimestamp and max MaxTimestamp from the parts which were merged. However, during the merge procedure VM check if data blocks within the part are already out of retention and will discard those which are.

**Comment by tantm3** [comment]
> Are you sure those gaps are caused by the retention and not because the metric was just absent at that time?

Yes, I am sure.
Because I create my own shell script to generate testing data.
Here is another scenario

![data](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/29f6ff88-24a9-483e-8813-128a59244016)

The removing part metrics is different between tags of the same metrics.
I also exported raw sample data here in case you needed it: 
[host_test_metrics.zip](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12060604/host_test_metrics.zip)

**Comment by tantm3** [comment]
I am sorry for the late response, too!
I try to create a procedure but I don't know where to start.

![retention-1](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/aa95f03f-fd58-4e34-897f-70a47772aaaa)

As you can see in the picture, I do a `last two days query` and it return a block of data on July 22.
I try to list the data parts:

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/d62e3a4a-040a-47a8-9c34-c6b634e7a551)

```
1917923_4781_20230722225038.039_20230723091344.370_176D953413569628
2702430_4805_20230722185824.000_20230723182013.858_176D95341356B0F9
995011_4774_20230722220418.157_20230723035530.414_176D953413568671
```
I think these parts are contained data from July 22.
So, is this a normal behavior from VM (retention 1d)?

> As you see, VM filters the samples out of retention on query time, while the data part is still on disk. If I force the merge - extra part will be deleted:
Is this work by default in the VM cluster?
Because of in the Grafana image above, I do the `last two days query` and VM still returns the data from July 22 which is out of retention 1 day.

**Comment by hagen1778** [comment]
Can you try updated version of VictoriaMetrics to perform the same test? Is it possible share snapshot of this data, so I can try locally? Can you reproduce this in isolated env via https://github.com/VictoriaMetrics/VictoriaMetrics/tree/master/deployment/docker ?

 > Is this work by default in the VM cluster or do we need to enable some configs?

Nothing except retentionPeriod cmd-line flag set on vmstorage nodes

**Comment by tantm3** [comment]
Just for information.
I am trying to deploy a new cluster with testing data and I will give you a snapshot data soon.

**Comment by tantm3** [comment]
Hi @hagen1778,

Here is the snapshot data of three vmstorage instances, I need to use Google Drive because of the max upload file size from Github, sorry for the inconvenience.
https://drive.google.com/drive/folders/1dlnVWrM1QBPVPamRSG2enOHwswbtIwNq?usp=sharing

Here is the image when I make a snapshot

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/0df7070a-cd9a-456a-9f92-c2c8215bf14a)

**Comment by hagen1778** [comment]
@tantm3 I believe the backup was done wrong. Archives listed at https://drive.google.com/drive/folders/1dlnVWrM1QBPVPamRSG2enOHwswbtIwNq?usp=sharing contain a lot of extra data not related to VM. And those folders which are related (indexdb, data) contains no data - only folder names. Are you able to restore from them?

**Comment by tantm3** [comment]
Hi @hagen1778,

Could you please try three new backup files?
I have a mistake when archiving the backup folder.

**Comment by hagen1778** [comment]
Sorry @tantm3, but I think there is still something wrong with provided files. Here's how I see it after downloading:
<img width="665" alt="image" src="https://github.com/VictoriaMetrics/VictoriaMetrics/assets/2902918/c485130a-4b5f-49e0-9b70-52cb44a07593">
All binary files are a folders now, and vmstorage can't start with this data.

I did a test on uploading a data folder to google drive:
1. stopped vmstorage
2. archived data folder
3. uploaded to google drive as file
4. downloaded it back 
5. was able to restore from the downloaded file - all files were of expected types.

Is your flow different from mine?

**Comment by tantm3** [comment]
Oh, I think the flow is different @hagen1778.
I do a snapshot and archived snapshot folder, not the raw data folder.
Here is the command I use:
```
/vmbackup-prod -storageDataPath=/storage -snapshot.createURL=http://localhost:8482/snapshot/create - -dst=fs:///opt/vmstorage
```
Could you please restore the snapshot or do you need the raw data folder?

**Comment by denisgolius** [comment]
> -dst=fs:///opt/vmstorage

Hi!

What filesystem use is used for `/opt/vmstorage` ? Is it NTFS from Windows or maybe you use NFS from some Windows OS?

**Comment by tantm3** [comment]
Hi,

I am sorry for the inconvenience.
I upload a new raw data folder and I am quite sure that the binary file is not become a folder anymore.
https://drive.google.com/drive/folders/1dlnVWrM1QBPVPamRSG2enOHwswbtIwNq?usp=sharing
I have no idea why the file becomes a folder.

**Comment by hagen1778** [comment]
Thanks! This worked!
I was able to run 3 vmstorage nodes on this data and I was able to see them (it was not easy to understand that data existed for 0:1 tenant, though). However, all the data parts I see on the mentioned backups is related to today or yesterday, 20230802 and 20230801 parts respectively. Given into account 1d retention - it is expected for them to be available for querying. But tomorrow, I shouldn't be able to see those data parts which cover 20230801 day only. So I'll check it tomorrow.

**Comment by tantm3** [comment]
>  (it was not easy to understand that data existed for 0:1 tenant, though)

I forgot to tell you the tenant id, I sent data to tenants 0:1 and 0:2. 
Could you please also check the scenario missing some blocks of data like in the image below?

![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/112939454/ea5cca95-b34a-40c9-8979-399e9c773ce3)

I look forward to seeing the result.

**Comment by hagen1778** [comment]
Hello! I was able to reproduce the issue today. I don't understand why it happens so far, but I'll get back to you once I do.

**Comment by tantm3** [comment]
Thanks @hagen1778 for your patience!
Feel free to ask me for the information.
I hope I could help you somewhat.

**Comment by hagen1778** [comment]
Hello @tantm3! Sorry for the late response.

> As you see, VM filters the samples out of retention on query time, while the data part is still on disk. If I force the merge - extra part will be deleted:

It turned out that this additional filtering on query time works on VictoriaMetrics single-server installation only, because in this case the process knows the configured retention period and is able to filter out results by time before returning.

In cluster mode, only vmstorage component is aware of retentionPeriod setting. This is why vmselect component isn't able to filter data in the same way as in single-server installation.

Taking into account that in both cases data parts still can be stored longer than configured retentionPeriod, the severity of this issue doesn't seem critical. Even if we fix that for cluster mode by, for example, explicitly specifying expected retentionPeriod for vmselect - it would only change the visual aspect. The data will remain on the disk until deleted.

I'd add that this issue might concerns installations with low retentionPeriod values, like `1d`. For bigger values of `retentionPeriod` it is unlikely a concern - this is why this issue was never a subject for discussion. Thanks for digging into details of this!

**Comment by tantm3** [comment]
Hi @hagen1778,

Thanks for your detailed response!
I think it's a good idea to keep the problem here as an enhancement.
Hope to see the enhancement soon!
